painter/src/painter.rs

```diff
@@ -17,61 +16,81 @@ use std::ops::{Deref, DerefMut};
 pub struct Painter {
-  size: Size,
+  bounds: Rect,
   state_stack: Vec<PainterState>,
   commands: Vec<PaintCommand>,
-  path_builder: Builder,
-  device_scale: f32,
+  path_builder: PathBuilder,
   typography_store: TypographyStore,
 }
 
-pub type CaptureCallback<'a> =
-  Box<dyn for<'r> FnOnce(DeviceSize, Box<dyn Iterator<Item = &[u8]> + 'r>) + 'a>;
-/// `PainterBackend` use to draw the picture what the `commands` described  to
-/// the target device. Usually is implemented by graphic library.
+/// `PainterBackend` use to draw textures for every frame, All `draw_commands`
+/// will called between `begin_frame` and `end_frame`
+///
+/// -- begin_frame()
+///
+///  +--> draw_commands --+
+///  ^                    V
+///  +----------<---------+
+///                       
+///
+///
+/// -+ end_frame()
+///                       
 pub trait PainterBackend {
-  /// Submit the paint commands to draw
-  fn submit(&mut self, commands: Vec<PaintCommand>);
+  type Texture;
+
+  fn set_anti_aliasing(&mut self, anti_aliasing: AntiAliasing);
+
+  /// A frame start.
+  fn begin_frame(&mut self);
 
-  fn commands_to_image(
+  /// Paint `commands` to the `output` Texture.
+  /// This may be called more than once during a frame.
+  fn draw_commands(
     &mut self,
+    viewport: DeviceRect,
     commands: Vec<PaintCommand>,
-    capture: CaptureCallback,
-  ) -> Result<(), Box<dyn Error>>;
+    surface: Color,
+    output: &mut Self::Texture,
+  );
 
-  fn resize(&mut self, size: DeviceSize);
+  /// A frame end.
+  fn end_frame(&mut self);
 }
 
-#[derive(Clone)]
-pub enum PaintPath {
-  Path(Resource<Path>),
-  Text {
-    font_size: FontSize,
-    glyphs: Vec<Glyph<Pixel>>,
-    style: PathStyle,
-  },
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub enum AntiAliasing {
+  None = 1,
+  Msaa2X = 2,
+  Msaa4X = 4,
+  Msaa8X = 8,
+  Msaa16X = 16,
 }
 
-// todo: need a way to batch commands as a single resource. so we can cache
-// their vertexes as a whole. useful for svg, animation and paint layers.
-#[derive(Clone)]
-pub struct PaintInstruct {
-  pub opacity: f32,
-  pub path: PaintPath,
+/// A path and its geometry information are friendly to paint and cache.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct PaintPath {
+  /// The path to painting, and its axis is relative to the `bounds`.
+  pub path: Path,
+  /// The bounds after path applied transform.
+  pub paint_bounds: Rect,
+  // The transform need to apply to the path.
   pub transform: Transform,
-  pub brush: Brush,
 }
 
-#[derive(Clone)]
-pub struct ClipInstruct {
-  pub path: PaintPath,
-  pub transform: Transform,
-}
-
-#[derive(Clone)]
+#[derive(Debug, Clone, Serialize, Deserialize)]
 pub enum PaintCommand {
-  Paint(PaintInstruct),
-  PushClip(ClipInstruct),
+  ColorPath {
+    path: PaintPath,
+    color: Color,
+  },
+  ImgPath {
+    path: PaintPath,
+    img: ShareResource<PixelImage>,
+    opacity: f32,
+  },
+  // Todo: keep rectangle clip.
+  Clip(PaintPath),
   PopClip,
```

Comment:
```suggestion
  PopClip,
```
here need more comment to describe it, I looked at the code for a while to know the meaning